### ISSUE TITLE
Demo feedback: Retention for stored notifications

### ISSUE THREAD

**Issue by Sephsekla** [WordPress/wp-feature-notifications#87]
Potential data storage issues were discussed in #83, however there was a specific focus on retention of notification data.

- Should this be limited to a specific number?
- Should dismissed notifications be purged after a certain retention period?
- Can notifications be superseded by a newer version? E.g. a plugin update notice is no longer relevant when a newer version becomes available.

Everyone was in agreement that wherever possible, these values should be filterable - see revisions for an example.

Relates to #56

**Comment by bacoords** [comment]
I've been wondering about this too. Marking as "read" clearly keeps something in the database, but should there be some (filterable) time limit to keeping this stuff, like how Trash has 30 days.

Similarly, is there going to be a UI for viewing ALL older notifications, even if read? Is that going in the hub? That seems like a lot of data to load.

**Comment by Sephsekla** [comment]
In theory loading all currently stored notifications shouldn't be a problem performance-wise if we're doing it asynchronously via REST. It can be paginated and loaded in batches.

That being said you're right that it's a lot, and I struggle to think of when it would be useful. Certainly a relatively conservative default that can be filtered would be good.

**Comment by bacoords** [comment]
Yeah good point about pagination. Trello has a similar panel for notifications and does an infinite load.

**Comment by erikyo** [comment]
I prepared something similar for a personal project that adds the infinite load (with a fetching to rest-api mechanism to receive data). I leave here [the link](https://github.com/erikyo/infinite-load-block/blob/7c7f768ac9098fdbdaf0f924b734affc82bf8b38/src/infinite-load-frontend.js#L17) because we could integrate a similar thing 

it's the implementation of this https://web.dev/patterns/web-vitals-patterns/infinite-scroll/infinite-scroll/ with React

---
> shouldn't be a problem performance-wise if we're doing it asynchronously via REST. It can be paginated and loaded in batches.

No, also because we can use localstorage to save user notifications so that we can use them as the redux initial state... that way there is no delay while we wait for the async response from the rest API.

**Comment by johnhooks** [comment]
In my opinion, there should be the option for a notification emitter to indicate whether or not a messages does expire and when, a database column of `expires_at` that is null or a timestamp. 

Deleting any message that is not flagged as expired, even if viewed or dismissed, would be unexpected behavior. Most types of messages might define an expiry, but if users begin to depend on a timeline of notifications that do not expire, perhaps for as part of a post edit/review workflow, plugin management, ... etc. It would be unexpected behavior if notifications were deleted without consent from the user or meeting the expiry criteria. 

I understand this means notifications will pile up. It will be important to store them effectively. Creating good documentation for plugin developers will be key to help educate them on best practices, and how to not overwhelm the users or their databases.
